File: backend/app/events.py
"""Load events from meet .lxf into the Event + AgeGroup tables."""
from __future__ import annotations

from pathlib import Path

from sqlalchemy.orm import Session
from .models import Event, AgeGroup
from .meet_parser import parse_meet_lxf, ParsedMeet


def _load_from_parsed(db: Session, meet: ParsedMeet) -> int:
    """Insert events + age groups from a parsed meet. Returns event count."""
    count = 0
    for ev in meet.all_events:
        event = Event(
            splash_event_id=ev.eventid,
            style_uid=ev.swimstyleid,
            style_name=ev.style_name or f"UID {ev.swimstyleid}",
            distance=ev.distance,
            relay_count=ev.relaycount,
            gender=ev.gender_int,
            event_number=ev.number,
            round=2 if ev.is_prelim else (1 if ev.round == "TIM" else 9),
            masters=ev.is_masters,
            session_id=None,
        )
        db.add(event)
        db.flush()
        for ag in ev.agegroups:
            db.add(AgeGroup(
                event_id=event.id,
                splash_agegroup_id=ag.agegroupid,
                age_min=ag.agemin,
                age_max=ag.agemax,
            ))
        count += 1
    db.commit()
    return count


def load_events(db: Session, lxf_path: Path) -> int:
    """Load events from meet .lxf if table is empty. Returns count."""
    if db.query(Event).first():
        return 0
    meet = parse_meet_lxf(lxf_path)
    return _load_from_parsed(db, meet)

File: backend/src/events.rs
use sqlx::PgPool;

use crate::meet_parser::ParsedMeet;

pub async fn load_from_parsed(pool: &PgPool, meet: &ParsedMeet) -> Result<i32, sqlx::Error> {
    let mut count = 0;
    for ev in meet.all_events() {
        let round_int = if ev.is_prelim() {
            2
        } else if ev.round == "TIM" {
            1
        } else {
            9
        };
        let row: (i32,) = sqlx::query_as(
            "INSERT INTO events (splash_event_id, style_uid, style_name, distance, relay_count, gender, event_number, round, masters, fee_cents)
             VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) RETURNING id"
        )
        .bind(ev.eventid)
        .bind(ev.swimstyleid)
        .bind(&ev.style_name)
        .bind(ev.distance)
        .bind(ev.relaycount)
        .bind(ev.gender_int())
        .bind(ev.number)
        .bind(round_int)
        .bind(ev.is_masters())
        .bind(ev.fee_cents)
        .fetch_one(pool)
        .await?;

        for ag in &ev.agegroups {
            sqlx::query(
                "INSERT INTO age_groups (event_id, splash_agegroup_id, age_min, age_max) VALUES ($1, $2, $3, $4)"
            )
            .bind(row.0)
            .bind(ag.agegroupid)
            .bind(ag.agemin)
            .bind(ag.agemax)
            .execute(pool)
            .await?;
        }
        count += 1;
    }
    Ok(count)
}

pub async fn load_events_if_empty(pool: &PgPool, lxf_path: &std::path::Path) -> Result<i32, String> {
    let exists: (i64,) = sqlx::query_as("SELECT COUNT(*) FROM events")
        .fetch_one(pool)
        .await
        .map_err(|e| e.to_string())?;
    if exists.0 > 0 {
        return Ok(0);
    }
    let data = std::fs::read(lxf_path).map_err(|e| e.to_string())?;
    let meet = crate::meet_parser::parse_meet_lxf(&data)?;
    load_from_parsed(pool, &meet).await.map_err(|e| e.to_string())
}
